src/server/deamon/index.ts

```diff
@@ -53,29 +59,35 @@ export class DaemonServer extends Dispose {
   constructor() {
     super();
     this.selectedDeviceId = this.config.get<string>(selectedDeviceIdKey);
-    this.push({
-      dispose: () => {
+    this.push(
+      Disposable.create(() => {
+        this.responseCallbacks.clear();
         try {
           this.process?.kill();
           this.process = undefined;
         } catch (error) {
           log(`dispose server error: ${error.message}`);
         }
-      },
-    });
+      }),
+    );
     this.eventHandlers['daemon.logMessage'] = this.logMessage;
     this.eventHandlers['daemon.connected'] = this.connected;
     this.eventHandlers['device.added'] = this.deviceAdded;
     this.eventHandlers['device.removed'] = this.deviceRemoved;
   }
 
-  async sendRequest(request: Request): Promise<void> {
-    if (!this.process || !this.process.stdin.writable) {
-      log(`Daemon not running but got request: ${JSON.stringify(request)}`);
-      return;
-    }
-    const rpcRequest = `[${JSON.stringify(request)}]\n`;
-    this.process.stdin.write(rpcRequest);
+  async sendRequest(request: Request): Promise<Message> {
+    return new Promise<Message>((res, rej) => {
+      if (!this.process || !this.process.stdin.writable) {
+        rej(`Daemon not running but got request: ${JSON.stringify(request)}`);
+        return;
+      }
+      this.responseCallbacks.set(request.id, (message: Message) => {
+        res(message);
```

Comment:
Might be nice to also add a timeout for the promise

Reply:
Done